spark-kafka-streaming-integration/src/main/java/org/thingsboard/samples/spark/KafkaStreamingDemoMain.java

```diff
@@ -0,0 +1,135 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.thingsboard.samples.spark;
+
+import lombok.extern.slf4j.Slf4j;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.common.serialization.StringDeserializer;
+import org.apache.spark.SparkConf;
+import org.apache.spark.streaming.Duration;
+import org.apache.spark.streaming.api.java.JavaInputDStream;
+import org.apache.spark.streaming.api.java.JavaStreamingContext;
+import org.apache.spark.streaming.kafka010.ConsumerStrategies;
+import org.apache.spark.streaming.kafka010.KafkaUtils;
+import org.apache.spark.streaming.kafka010.LocationStrategies;
+import org.eclipse.paho.client.mqttv3.*;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.OptionalDouble;
+
+
+public class KafkaStreamingDemoMain {
+
+    public static void main(String[] args) throws Exception {
+        new StreamRunner().start();
+    }
+
+    @Slf4j
+    private static class StreamRunner {
+
+        private static final String AVERAGE_DEVICE_ACCESS_TOKEN = "h8mLQPqbYOot6F8kz1oj";
+        private static final String KAFKA_BROKER_LIST = "localhost:9092";
+        private static final String THINGSBOARD_MQTT_ENDPOINT = "tcp://localhost:1883";
+        private static final int STREAM_WINDOW_MILLISECONDS = 10000; // 10 seconds
+        private final MqttAsyncClient client;
+        private Collection<String> topics = Arrays.asList("sensors-telemetry");
+
+        StreamRunner() throws MqttException {
+            client = new MqttAsyncClient(THINGSBOARD_MQTT_ENDPOINT, MqttAsyncClient.generateClientId());
+        }
+
+        void start() throws Exception {
+            SparkConf conf = new SparkConf().setAppName("Kafka Streaming App").setMaster("local[2]");
+
+            try (JavaStreamingContext ssc = new JavaStreamingContext(conf, new Duration(STREAM_WINDOW_MILLISECONDS))) {
+
+                connectToThingsboard();
+
+                JavaInputDStream<ConsumerRecord<String, String>> stream =
+                        KafkaUtils.createDirectStream(
+                                ssc,
+                                LocationStrategies.PreferConsistent(),
+                                ConsumerStrategies.<String, String>Subscribe(topics, getKafkaParams())
+                        );
+
+                stream.foreachRDD(rdd ->
+                {
+                    OptionalDouble averageTemp = rdd.mapToDouble(r ->
```

Comment:
You should first reduce the rdd to Pair<Sum, Count> and only then collect one record. Using that record you can calculate average. Otherwise all heavy work happens in the driver and not in the worker.